nifi-extension-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java

```diff
@@ -302,8 +302,14 @@ String resolvePythonCommand() throws IOException {
         } else if (virtualEnvDirectories.length == 1) {
             commandExecutableDirectory = virtualEnvDirectories[0].getName();
         } else {
-            // Default to bin directory for macOS and Linux
-            commandExecutableDirectory = "bin";
+            // We should get at most 2 directories.  Check for python command.
+            if (virtualEnvDirectories[0].list((file, name) -> name.startsWith(pythonCmd)).length >= 1) {
```

Comment:
The `list()` method can return `null` under some circumstances, so adjusting the approach seems necessary. What about moving this block into a separate method named something like `findExecutableDirectory`? That method could iterate over the `virtualEnvDirectories` array, and not necessary assume that there will be at most two directories.

Reply:
Yes I agree it is safer if I make it null safe and cleaner with a new method.  The challenge was working with the virtualEnvDirectories array in a for loop and ensuring commandExecutableDirectory is only set exactly once.  If I break I don't guarantee it's set if I loop without breaking it may be set multiple.  So I've replaced this with a stream which is not the most efficient but  does make the code readable and keeps the final guarantee or throws an exception.